Strict filtering on barcode 2,5 and 8 from cvn1

```{r setup}

#install.packages(tidyverse)

library(tidyverse)


```



Load in the taxa file for all barcodes (previously filtered in assigning_taxonomy.Rmd to remove duplicate reads with distinct function. No unclassified reads or misassigned barcodes are in this file either. 19/03/2025)
```{r}
load("run1_culturevsnan_taxa_distinct")

```



```{r}

cvn1_stringent_2025_clean <- run1_culturevsnan_taxa_distinct %>%
                mutate(proportional_gap_number = gaps/alignment.length) %>%# 7.11.24 adding this as i think gaps set at less than 5 is too harsh for longer reads 
                mutate(proportional_mismatch_number = mismatches/alignment.length) %>% # 7.11.24 adding this as i think gaps set at less than 5 is too harsh for longer reads 
 # filter(gaps <= 5) %>% #and across the whole dataset, filter out any reads that have gaps more than 5 and
               #filter(mismatches <= 200) %>% #mismatches more than 200
              
              filter(proportional_mismatch_number <= 0.1) %>%
              filter(proportional_gap_number <= 0.1) %>% #here i am saying that there can be up to and including 10% gaps in the alignment 
                          #l ine 428-433 is general filtering of the entire dataset
               # filter(mismatches <= 200) %>%
                mutate(alignment.prop = alignment.length/query.length)%>%
                filter(alignment.prop >= 0.25) %>%
                filter(query.length >= 500)%>%
                filter(subject.length >= 100)#%>% #line 428-433 is general filtering of the entire dataset
  

cvn1_stringent_June2025_clean <- cvn1_stringent_2025_clean %>% 
               group_by (query.acc, subject.tax.ids) %>% # code from line 436-451 written with nick as the weighted approach 25.10.24
            
               mutate(evalue_nonzero = ifelse(evalue == 0, 1, evalue), 
                      relative_evalue = ifelse(evalue == 0, 1, min(evalue_nonzero)/evalue_nonzero)) %>%
              mutate(relative_alignment_length = alignment.length/max(alignment.length)) %>%
              mutate(alignment_count = n()) %>% #taxon alignment count within a read 
              
              ungroup() %>%
              group_by(query.acc) %>%
              filter(!species.y == "NA") %>% 
              mutate(relative_count = alignment_count/n()) %>% #proportion of read alignments to a taxon
               mutate(weight_score = relative_evalue*relative_alignment_length*(percent.identity/100)) %>%
             filter(weight_score == max(weight_score)) %>%
              filter(relative_count == max(relative_count)) %>%
              slice_sample(n=1) %>%
              ungroup()

save(cvn1_stringent_June2025_clean, file = "cvn1_stringent_June2025_clean") #saving the file for quick reloading later on

load("cvn1_stringent_June2025_clean")

#time took = 5 mins, ran fine the first time
```

```{r}

#need to add in sample name, time and location to cvn1 data plus make the prop table for use in bacterial plots:

# CVN1 samples to add: barcode 2,5 and 8. In the directory (C:\Users\keegk\moredungroup\Voyager Online - Nanopore zoonotic disease AMR PhD Files\Move Files Here\SQKRBK114.24\culturing_vs_nanopore\Culture_vs_Nanopore_1) there is a file called "run1_positive_cultures" which just contain barcode 2,5 and 8 after strict filtering (made in the R notebook run1_pos_neg_cultures.Rmd from the Culture_vs_Nanopore_1 R project. ) Loading this file  here:
 
load("cvn1_stringent_June2025_clean")


##below is example code, need to fill in cvn1 data :
  #ADD SAMPLE NAMES
cvn1_stringent_2025_clean3 <- cvn1_stringent_June2025_clean %>%
  
mutate(sample_name = case_when(
 barcode == "barcode1" ~ "FB1_25mg",
 barcode == "barcode2" ~ "FB1", # this is FB1_200, but im renaming it just FB1, as all my sample in this chapter are from 200 mg of faeces and i dont want FB1_200 in any plots
 barcode == "barcode3" ~ "FB1_noprec",
barcode == "barcode4" ~ "FC2_25mg",
barcode == "barcode5" ~ "FC2",# this is Fc2_200, but im renaming it just FB1, as all my sample in this chapter are from 200 mg of faeces and i dont want FB1_200 in any plots
 barcode == "barcode6" ~ "FC2_noprec",
 barcode == "barcode7" ~ "FE08_25mg",
 barcode == "barcode8" ~ "FE08",# this is Fe08_200, but im renaming it just FB1, as all my sample in this chapter are from 200 mg of faeces and i dont want FB1_200 in any plots
barcode == "barcode9" ~ "FE08_noprec",
 barcode == "barcode10" ~ "WG14_10^8",
 barcode == "barcode11" ~ "WG14_10^2",
 barcode == "barcode12" ~ "WG14_10^1",
barcode == "barcode13" ~ "WG14_10^8_noprec",
 barcode == "barcode14" ~ "WG14_10^2_noprec",
barcode == "barcode15" ~ "WG14_10^1_noprec",
barcode == "barcode16" ~ "WG14_10^8_200",
 barcode == "barcode17" ~ "WG14_10^2_200",
 barcode == "barcode18" ~ "WG14_10^1_200",
 barcode == "barcode19" ~ "WG14_10^8_200_noprec",
barcode == "barcode20" ~ "WG14_10^2_200_noprec",
barcode == "barcode21" ~ "WG14_10^1_200_noprec",
barcode == "barcode22" ~ "WG14_25mg",
barcode == "barcode23" ~ "HMW_POS_CTRL",
barcode == "barcode24" ~ "LOG_POS_CTRL",
TRUE ~ NA_character_
 )) %>%
  
    #ADD lOCATION 
mutate(location = case_when(
 barcode == "barcode1" ~ "Orkney",
 barcode == "barcode2" ~ "Orkney",
 barcode == "barcode3" ~ "Orkney",
barcode == "barcode4" ~ "Orkney",
 barcode == "barcode5" ~ "Orkney",
 barcode == "barcode6" ~ "Orkney",
 barcode == "barcode7" ~ "Orkney",
 barcode == "barcode8" ~ "Orkney",
barcode == "barcode9" ~ "Orkney",
 barcode == "barcode10" ~ "Orkney",
 barcode == "barcode11" ~ "Orkney",
 barcode == "barcode12" ~ "Orkney",
barcode == "barcode13" ~ "Orkney",
 barcode == "barcode14" ~ "Orkney",
barcode == "barcode15" ~ "Orkney",
barcode == "barcode16" ~ "Orkney",
 barcode == "barcode17" ~ "Orkney",
 barcode == "barcode18" ~ "Orkney",
 barcode == "barcode19" ~ "Orkney",
barcode == "barcode20" ~ "Orkney",
barcode == "barcode21" ~ "Orkney",
barcode == "barcode22" ~ "Orkney",
barcode == "barcode23" ~ "NA",
barcode == "barcode24" ~ "NA",
TRUE ~ NA_character_
 ))%>%

  #ADD time

mutate(time = case_when(
 barcode == "barcode1" ~ "November_22",
 barcode == "barcode2" ~ "November_22",
 barcode == "barcode3" ~ "November_22",
barcode == "barcode4" ~ "November_22",
 barcode == "barcode5" ~ "November_22",
 barcode == "barcode6" ~ "November_22",
 barcode == "barcode7" ~ "November_22",
 barcode == "barcode8" ~ "November_22",
barcode == "barcode9" ~ "November_22",
 barcode == "barcode10" ~ "June_23",
 barcode == "barcode11" ~ "June_23",
 barcode == "barcode12" ~ "June_23",
barcode == "barcode13" ~ "June_23",
 barcode == "barcode14" ~ "June_23",
barcode == "barcode15" ~ "June_23",
barcode == "barcode16" ~ "June_23",
 barcode == "barcode17" ~ "June_23",
 barcode == "barcode18" ~ "June_23",
 barcode == "barcode19" ~ "June_23",
barcode == "barcode20" ~ "June_23",
barcode == "barcode21" ~ "June_23",
barcode == "barcode22" ~ "June_23",
barcode == "barcode23" ~ "NA",
barcode == "barcode24" ~ "NA",
TRUE ~ NA_character_
 ))%>%
  
    #ADD RUN NAME

mutate(run = case_when(
 barcode == "barcode1" ~ "cvn1",
 barcode == "barcode2" ~ "cvn1",
 barcode == "barcode3" ~ "cvn1",
barcode == "barcode4" ~ "cvn1",
 barcode == "barcode5" ~ "cvn1",
 barcode == "barcode6" ~ "cvn1",
 barcode == "barcode7" ~ "cvn1",
 barcode == "barcode8" ~ "cvn1",
barcode == "barcode9" ~ "cvn1",
 barcode == "barcode10" ~ "cvn1",
 barcode == "barcode11" ~ "cvn1",
 barcode == "barcode12" ~ "cvn1",
barcode == "barcode13" ~ "cvn1",
 barcode == "barcode14" ~ "cvn1",
barcode == "barcode15" ~ "cvn1",
barcode == "barcode16" ~ "cvn1",
 barcode == "barcode17" ~ "cvn1",
 barcode == "barcode18" ~ "cvn1",
 barcode == "barcode19" ~ "cvn1",
barcode == "barcode20" ~ "cvn1",
barcode == "barcode21" ~ "cvn1",
barcode == "barcode22" ~ "cvn1",
barcode == "barcode23" ~ "cvn1",
barcode == "barcode24" ~ "cvn1",
TRUE ~ NA_character_
 ))

save(cvn1_stringent_2025_clean3, file = "cvn1_stringent_2025_clean3") 
load("cvn1_stringent_2025_clean3")
   # group_by(phylum, sample_name) %>%

 cvn1_stringent_2025_clean4 <-   cvn1_stringent_2025_clean3 %>%
   filter(barcode %in% c("barcode2","barcode5","barcode8")) # %>% #removing the controls and spiked campy samples
 save(cvn1_stringent_2025_clean4, file = "cvn1_stringent_2025_clean4") 
load("cvn1_stringent_2025_clean4") 

 stringent_2025_cvn1_prop  <- cvn1_stringent_2025_clean4  %>% 
  filter(superkingdom == "Bacteria") %>%
  group_by(sample_name, phylum) %>%
  distinct(subject.tax.ids) %>% #adding this 26.4.24 to remove multiple hits to the same taxa, inflating the numbers
  summarise(phylum_count = n()) %>%
  mutate(prop =prop.table(phylum_count)) %>%
  ungroup() %>%
  arrange(desc(prop)) %>%
  mutate(phylum = replace_na(phylum, "Unassigned")) %>%
  mutate(phylum = factor(phylum, levels = unique(phylum)))
 
 

save(stringent_2025_cvn1_prop, file = "stringent_2025_cvn1_prop") #saving the file for quick reloading later on

##Genus prop 
  
 stringent_2025_cvn1_prop_genus  <- cvn1_stringent_2025_clean3 %>% 
  filter(superkingdom == "Bacteria") %>%
  group_by(sample_name, genus) %>%
  distinct(subject.tax.ids) %>% #adding this 26.4.24 to remove multiple hits to the same taxa, inflating the numbers
  summarise(genus_count = n()) %>%
  mutate(prop =prop.table(genus_count)) %>%
  ungroup() %>%
  arrange(desc(prop)) %>%
  mutate(genus = replace_na(genus, "Unassigned")) %>%
  mutate(genus = factor(genus, levels = unique(genus)))

save(stringent_2025_cvn1_prop_genus, file = "stringent_2025_cvn1_prop_genus") #saving the file for quick reloading later on

##family prop 
  
 stringent_2025_cvn1_prop_family  <-cvn1_stringent_2025_clean3 %>% 
  filter(superkingdom == "Bacteria") %>%
  group_by(sample_name, family) %>%
  distinct(subject.tax.ids) %>% #adding this 26.4.24 to remove multiple hits to the same taxa, inflating the numbers
  summarise(family_count = n()) %>%
  mutate(prop =prop.table(family_count)) %>%
  ungroup() %>%
  arrange(desc(prop)) %>%
  mutate(family = replace_na(family, "Unassigned")) %>%
  mutate(family = factor(family, levels = unique(family)))

save(stringent_2025_cvn1_prop_family , file = "stringent_2025_cvn1_prop_family") #saving the file for quick reloading later on



```